### PR TITLE
feat(eslint-plugin): add quotes [extension]

### LINKED ISSUE #756: [quotes] False positive requiring backticks in types
**Repro**

```JSON
{
  "rules": {
    "quotes": ["error", "backtick", { "avoidEscape": true }]
  }
}
```

```TS
class MockObserver {
  public hasSubscribers: IBindingTargetObserver['hasSubscribers'];
}
```

**Expected Result**
No rule violations as those can't be backticks.

**Actual Result**
Violation of `quotes`.

**Additional Info**

Actually changing them to backticks gives a compiler error.
```
public hasSubscribers: IBindingTargetObserver[`hasSubscribers`];
```
Error: `']' expected`

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.12.0` |
| `@typescript-eslint/parser`        | `1.12.0` |
| `TypeScript`                       | `3.5.3` |
| `ESLint`                           | `5.16.0` |
| `node`                             | `10.15.0` |
| `npm`                              | `6.10.2` |

### LINKED ISSUE #757: [quotes] False positive with backtick in declare module
**Repro**

```JSON
{
  "rules": {
    "quotes": ["error", "backtick", { "avoidEscape": true }]
  }
}
```

`html.d.ts`:
```TS
declare module '*.html' {
  const value: string;
  export default value;
}
```

**Expected Result**
No rule violations as those can't be backticks.

**Actual Result**
Violation of `quotes`.

**Additional Info**

Actually changing them to backticks gives a compiler error.
```
declare module `*.html` {
  const value: string;
  export default value;
}
```
Error: `Cannot find name 'declare'`

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.12.0` |
| `@typescript-eslint/parser`        | `1.12.0` |
| `TypeScript`                       | `3.5.3` |
| `ESLint`                           | `5.16.0` |
| `node`                             | `10.15.0` |
| `npm`                              | `6.10.2` |

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #757
Fixes #756

### CODE REVIEW

File: packages/eslint-plugin/src/rules/quotes.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    };
+
+    const isTypeLiteral = (node: TSESTree.Literal): boolean => {
+      return !!(
+        node.parent && node.parent.type === AST_NODE_TYPES.TSLiteralType
+      );
+    };
+
+    return {
+      Literal(node: TSESTree.Literal) {
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/quotes.ts]: this should be automatically typed for you
```suggestion
      Literal(node) {
```

File: packages/eslint-plugin/src/rules/quotes.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+          option === 'backtick' &&
+          (isModuleDeclaration(node) || isTypeLiteral(node))
+        ) {
+          return;
+        }
+
+        rules.Literal(node);
+      },
+
+      TemplateLiteral(node: TSESTree.TemplateLiteral) {
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/quotes.ts]: ditto
```suggestion
      TemplateLiteral(node) {
```

File: packages/eslint-plugin/src/rules/quotes.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    {
+      allowTemplateLiterals: false,
+      avoidEscape: false,
+    },
+  ],
+  create(context, [option]) {
+    const rules = baseRule.create(context);
+
+    const isModuleDeclaration = (node: TSESTree.Literal): boolean => {
+      return !!(
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/quotes.ts]: you should move the `!!` so it's next to the `node.parent` instead

File: packages/eslint-plugin/src/rules/quotes.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        node.parent && node.parent.type === AST_NODE_TYPES.TSModuleDeclaration
+      );
+    };
+
+    const isTypeLiteral = (node: TSESTree.Literal): boolean => {
+      return !!(
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/quotes.ts]: ditto

**Comment by a-tarasyuk** [comment]
@bradzacher added changes.

**Comment by LekoArts** [comment]
Would be great to have that merged! Just stumbled upon the initial issue

**Comment by regevbr** [comment]
Hi @JamesHenry @a-tarasyuk 
Thanks for the fix, but there is 1 more case which is not covered here.
When you need to import from old modules that use the following syntax
```typescript
export = ...
```
You must use
```typescript
import moment = require('moment'); // real life example
```
to import them, and the rule alerts on those, while it is not valid to use back ticks here.

Do you can care to fix that as well?

**Comment by bradzacher** [comment]
hey @regevbr - please raise a new issue instead of commenting on old PRs.

**Pull request merged**